### ISSUE TITLE
Many warnings with gcc 11

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#2229]
If compiled with gcc 11, there are many warnings like:
```
../../include/bout/../msg_stack.hxx: In destructor 'MsgStackItem::~MsgStackItem()':
../../include/bout/../msg_stack.hxx:147:15: warning: 'bool std::uncaught_exception()' is deprecated [-Wdeprecated-declarations]
  147 |     if (!std::uncaught_exception()) {
      |               ^~~~~~~~~~~~~~~~~~
```
I think is should be updated to `std::uncaught_exceptions()`

**Comment by ZedThree** [comment]
This is because gcc 11 uses `-std=c++17` by default now, and `std::uncaught_exception` was deprecated in C++17. Unfortunately, `std::uncaught_exceptions()` was also only added in C++17. Easiest thing for the time being might be to explicitly set the used standard at configure time via with `CMAKE_CXX_FLAGS` or `CXX_FLAGS`.

Obviously we will need to fix this properly at some point! I'm not 100% certain that `uncaught_exceptions` is a drop-in replacement.

We starting to build up a few backports, might be nice to move them from `utils.hxx` to something like `std_backports.hxx`.

**Comment by ZedThree** [comment]
There's a fix in #2274, but I'm happy pull that change into a separate PR

**Comment by ZedThree** [comment]
Fixed in #2276
